Grdowns/build tasks

grdowns opened the pull request:
Generate build tasks for the user which will compile the active file and output a binary in the same directory
The build tasks are generated based off of compilers found on the user's machine, as well as their compilerPath setting

grdowns committed:
Generate a C/C++ build task

grdowns committed:
Cleaning

grdowns committed:
Move args into definition

grdowns committed:
Add args to shellexecution to allow calling the task without first configuring; Add TODO's

grdowns committed:
Read compiler paths sent in defaults to generate task list

grdowns committed:
Rename compilerPaths to compilerInfo; Filter tasks based on file extension

grdowns committed:
Use our own file extension filters to decide whether tasks are presented

grdowns committed:
Show dialog when no compilers are found

grdowns committed:
Display a message if no compilers are found

grdowns committed:
Add compilerPath getter to config; Use it to offer another task

grdowns committed:
Refactoring; Commenting

grdowns committed:
Remove CompilerInfo namespace; Formatting; Change return vals

grdowns committed:
Commenting; Add telemetry; Inverse condition

grdowns committed:
Change output program based on OS; Remove message and collect telemetry instead

grdowns committed:
Filter out redundant compiler entries

grdowns committed:
Remove TODO; Get proper exe name; Properly reassign compilerPaths after filtering

grdowns committed:
Commenting; Fix cwd mingw bug

grdowns committed:
Commenting; Refactor

grdowns committed:
Mark compilerInfo const

grdowns committed:
Remove whitespace; Remove bad comment

grdowns committed:
Remove whitespace

grdowns committed:
Replace unique algorithm with a map

grdowns committed:
pre-compute basename

pieandcakes commented on Extension/src/LanguageServer/extension.ts:
```diff
@@ -82,6 +99,113 @@ export function activate(activationEventOccurred: boolean): void {
+        // const learnMore: string = "Learn More";
+        // const message: string = "No C/C++ compiler found on the system. Please install a C/C++ compiler to use the C/Cpp: build active file tasks.";
+
+        // let showNoCompilerFoundMessage: PersistentState<boolean> = new PersistentState<boolean>("CPP.showNoCompilerFoundMessage", true);
```
is this commented out for a reason? or is that what that first line of the comment is explaining? If so, i'd put it on a new line so its less confusing

pieandcakes commented on Extension/src/LanguageServer/extension.ts:
```diff
@@ -82,6 +99,113 @@ export function activate(activationEventOccurred: boolean): void {
+    let result: vscode.Task[] = [];
+    compilerPaths.forEach(compilerPath => {
+        const taskName: string = path.basename(compilerPath) + " build active file";
+        const args: string[] = ['-g', '${fileDirname}/${fileBasename}', '-o', '${fileDirname}/' + exeName];
```
To avoid accidental collisions, should the output exes be put in a \bin folder inside the directory?

grdowns replied:
Not a bad idea. Another thought is putting it into the `.vscode` folder. @bobbrow?

pieandcakes replied:
I would be against `.vscode` as that seems to be where configuration and extension data is going and not anything to do with the customer's coding experience. For me it would be between `/bin` or `/out`.

pieandcakes commented on Extension/src/LanguageServer/extension.ts:
```diff
@@ -82,6 +99,113 @@ export function activate(activationEventOccurred: boolean): void {
+    let result: vscode.Task[] = [];
+    compilerPaths.forEach(compilerPath => {
+        const taskName: string = path.basename(compilerPath) + " build active file";
+        const args: string[] = ['-g', '${fileDirname}/${fileBasename}', '-o', '${fileDirname}/' + exeName];
```
Also, to be easier on the user, could we do `<command> -g *.<whatever their file extension choice is such as cpp> -o <filename>.<exe>` ?

grdowns replied:
As in attempt to compile all files within the current working directory? And how would they select the file extension of choice through the quickpick menu? I'm not sure I understand

pieandcakes replied:
Based on the file extension of the currently opened file? Or chain all of them? I don't imagine beginning users who would use this to use .c .cpp and .cxx all together in the same project.

grdowns committed:
Move comment

grdowns committed:
Commenting; Ensure user's compilerPath setting is used

grdowns committed:
Use indexOf instead of find

grdowns committed:
Simplify isHeader calculation by using some instead of every

grdowns committed:
Don't redundantly check extensions; Use some instead of indexOf

grdowns committed:
Merge branch 'master' into grdowns/build-templates

sean-mcmanus committed:
Merge branch 'master' into grdowns/build-templates

sean-mcmanus commented on Extension/src/LanguageServer/extension.ts:
```diff
@@ -83,6 +100,117 @@ export function activate(activationEventOccurred: boolean): void {
+        const command: vscode.ShellExecution = new vscode.ShellExecution(compilerPath, args, { cwd: path.dirname(compilerPath) });
+        const target: vscode.WorkspaceFolder = vscode.workspace.getWorkspaceFolder(clients.ActiveClient.RootUri);
+        let task: vscode.Task = new vscode.Task(kind, target, taskName, 'C/Cpp', command, '$gcc');
+        task.definition = kind; // The constructor for vscode.Task will eat the definition. Reset it by reassigning
```
I'd prefer more usage of periods at the end of comments...not sure if we're enforcing that.

grdowns committed:
Add comment punctuation

Use file macro instead of fileDirName and fileBaseName

Change path provider name; expose cwd in definition

Add shell task definition (it's not built into vs code apparently)

grdowns committed:
Merge branch 'grdowns/build-templates' of https://github.com/Microsoft/vscode-cpptools into grdowns/build-templates

sean-mcmanus commented on Extension/src/LanguageServer/extension.ts:
```diff
@@ -83,6 +100,119 @@ export function activate(activationEventOccurred: boolean): void {
+    }
+
+    // Don't offer tasks for header files.
+    const activeFileExt: string = path.extname(activeEditor.document.fileName);
```
This code is case-sensitive, right? It seems like it needs to handle both cases then. For example, .H should count as a header file too, and .C is supposed to be a C++ file.

Also, extensionless files should count as header files, if that's possible.

grdowns committed:
Lowercase file extension to match; Count extensionless files as headers

grdowns committed:
Merge branch 'master' into grdowns/build-templates

grdowns committed:
Add newline

grdowns committed:
Merge branch 'grdowns/build-templates' of https://github.com/Microsoft/vscode-cpptools into grdowns/build-templates

bobbrow commented on Extension/src/LanguageServer/extension.ts:
```diff
@@ -83,6 +100,119 @@ export function activate(activationEventOccurred: boolean): void {
+    }
+
+    // Don't offer tasks if the active file's extension is not a recognized C/C++ extension.
+    const activeFileIsCpp: boolean = [".cpp", ".cc", ".cxx", ".mm", ".ino", ".inl"].some(ext => activeFileExt === ext);
```
if this is case-sensitive, consider dealing with `.C` (which is c++, by convention).  Perhaps we should offer both in this case.

bobbrow commented on Extension/src/LanguageServer/configurations.ts:
```diff
@@ -46,6 +46,7 @@ export interface ConfigurationJson {
 export interface Configuration {
     name: string;
     compilerPath?: string;
+    compilerInfo?: CompilerInfo[];
```
I usually use plurals in the variable names for arrays/lists.  e.g. consider renaming to `'compilers` or `knownCompilers` (`compilerInfos` sounds weird)

grdowns committed:
Change CompilerInfo to KnownCompiler

grdowns committed:
Support ambiguous file extension

grdowns committed:
Merge branch 'master' into grdowns/build-templates

grdowns committed:
Early out for extensionless files

grdowns committed:
Merge branch 'grdowns/build-templates' of https://github.com/Microsoft/vscode-cpptools into grdowns/build-templates

grdowns committed:
Commenting

bobbrow commented on Extension/src/LanguageServer/extension.ts:
```diff
@@ -128,7 +128,7 @@ async function getBuildTasks(): Promise<vscode.Task[]> {
     } else if (activeFileExt === '.c') {
         activeFileLanguage = 'c';
-    } else if (activeFileExt !== '.C') { // Ambiguous file extension
+    } else if (activeFileExt !== '.C') { // Ambiguous file extension.
```
ambiguous means show tasks for both c and c++

grdowns replied:
I updated the comment to reflect that's the case

bobbrow commented on Extension/src/LanguageServer/extension.ts:
```diff
@@ -83,6 +100,128 @@ export function activate(activationEventOccurred: boolean): void {
+
+    if (!compilerPaths) {
+        // Don't prompt a message yet until we can make a data-based decision.
+        telemetry.logLanguageServerEvent('noCompilerFound');
```
can we call it "buildTaskNoCompiler"?

grdowns committed:
Commenting; Rename telemetry event

grdowns committed:
Change languageAssociation to isC

grdowns committed:
Fix bug where cwd was not used; Add comment
